x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizer.java

```diff
@@ -427,4 +434,43 @@ public static boolean isPushableFieldAttribute(Expression exp, Predicate<FieldAt
         return false;
     }
 
+    private static class SpatialFromDocValuesExtraction extends OptimizerRule<AggregateExec> {
+        @Override
+        protected PhysicalPlan rule(AggregateExec aggregate) {
+            var foundAttribute = new Holder<Attribute>(null);
+
+            PhysicalPlan plan = aggregate.transformDown(UnaryExec.class, exec -> {
+                if (exec instanceof AggregateExec agg) {
+                    var orderedAggregates = new ArrayList<NamedExpression>();
+                    var changedAggregates = false;
+                    for (NamedExpression aggExpr : agg.aggregates()) {
+                        if (aggExpr instanceof Alias as && as.child() instanceof SpatialAggregateFunction af) {
+                            // Firstly we tell the aggregation function that is should expect doc-values instead of source values
+                            var changed = as.replaceChild(af.withDocValues());
+                            changedAggregates = true;
+                            if (af.field() instanceof Attribute fieldAttribute) {
+                                foundAttribute.set(fieldAttribute);
+                            } else {
+                                throw new RuntimeException("Expected field attribute, got " + af.field());
+                            }
```

Comment:
I added tests. However, while the following works:

```
row wkt = "POINT(42.97109629958868 14.7552534006536)"
| STATS c = ST_CENTROID(TO_GEOPOINT(wkt));
```

Doing the string literal inline does not:

```
row dummy = "dummy"
| STATS c = ST_CENTROID(TO_GEOPOINT("POINT(42.97109629958868 14.7552534006536)"));
```

Yeilds:

```
org.elasticsearch.xpack.ql.InvalidArgumentException:
  Does not support yet aggregations over constants
  - [ST_CENTROID(TO_GEOPOINT("POINT(42.97109629958868 14.7552534006536)"))]
```

That exception is throw by code in `AbstractPhysicalOperationProviders` that includes the comment:

```
// Only count can now support literals - all the other aggs should be optimized away
```

So it seems this is not yet supported.

Reply:
Anyway, use of fields that are reference fields, not literal fields, does not even reach the local optimization because no `FieldExtractExec` is produced, and therefor not optimized. I added a test to the `PhysicalOptimizationTests` to verify this.